File: vocabulary-command-service/src/main/java/com/learnwords/vocabularycommandservice/entity/Outbox.java
package com.learnwords.vocabularycommandservice.entity;


import com.learnwords.common.AggregateType;
import com.learnwords.common.EventStatus;
import com.learnwords.common.EventType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;

import lombok.NoArgsConstructor;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.Instant;


@AllArgsConstructor
@NoArgsConstructor
@Builder
@Entity
@Table(name = "outbox",
        indexes = {
                @Index(name = "idx_outbox_status", columnList = "eventStatus"),
                @Index(name = "idx_outbox_created", columnList = "createdAt")
        })
public class Outbox {
    @Id
    @Column(length = 36, updatable = false)
    private String eventId;

    @Column(nullable = false, length = 36)
    private String aggregateId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    private AggregateType aggregateType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private EventType eventType;

    @Column(columnDefinition = "jsonb", nullable = false)
    @JdbcTypeCode(SqlTypes.JSON)
    private String payload;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    @Builder.Default
    private EventStatus eventStatus = EventStatus.CREATED;

    @Column(nullable = false)
    @Builder.Default
    private int retryCount = 0;

    @Column(length = 36, nullable = true)
    private String deckId;

    @Column(nullable = false, updatable = false)
    @Builder.Default
    private Instant createdAt = Instant.now();

    @Column(nullable = false)
    private Instant updatedAt;

    @PrePersist
    public void prePersist() {
        updatedAt = createdAt;
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = Instant.now();
    }
}

File: koog-service/src/main/kotlin/com/learnwords/koogservice/persistence/entity/Outbox.kt
package com.learnwords.koogservice.persistence.entity

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.Table
import org.hibernate.annotations.JdbcTypeCode
import org.hibernate.type.SqlTypes
import java.time.Instant

/**
 * Encja Outbox do wzorca transactional outbox.
 * 
 * Przechowuje eventy do wysłania przez Kafka.
 */
@Entity
@Table(name = "koog_outbox")
class Outbox(
    @Id
    @Column(name = "event_id", nullable = false, unique = true)
    val eventId: String? = null,
    
    @Column(name = "aggregate_type", nullable = false)
    val agregateType: String = "",
    
    @Column(name = "aggregate_id", nullable = false)
    val agregateId: String = "",
    
    @Column(name = "event_type", nullable = false)
    val eventType: String = "",
    
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "payload", columnDefinition = "jsonb")
    val payload: String? = null,
    
    @Column(name = "status", nullable = false)
    var status: String = "PENDING",
    
    @Column(name = "retry_count", nullable = false)
    var retryCount: Int = 0,
    
    @Column(name = "created_at", nullable = false)
    val createdAt: Instant = Instant.now(),
    
    @Column(name = "updated_at", nullable = false)
    var updatedAt: Instant = Instant.now()
) {
    protected constructor() : this(eventId = null)
}